### PR TITLE
expression: implement vectorized evaluation for `builtinCastIntAsDurationSig` 

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
Implement vectorized evaluation for `builtinCastIntAsDurationSig`.

### What is changed and how it works?
10% faster than before:
```
BenchmarkVectorizedBuiltinCastFunc/builtinCastIntAsDurationSig-VecBuiltinFunc-12           30000             45574 ns/op               0 B/op          0 allocs/op
BenchmarkVectorizedBuiltinCastFunc/builtinCastIntAsDurationSig-NonVecBuiltinFunc-12        30000             51745 ns/op               0 B/op          0 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/bench_test.go [hunk 1]
```diff
@@ -574,13 +581,12 @@ func (s *testEvaluatorSuite) TestVectorizedBuiltinFunc(c *C) {
...
 				err := baseFunc.vecEvalDuration(input, output)
 				c.Assert(err, IsNil)
-				d64s := output.GoDurations()
 				for row := it.Begin(); row != it.End(); row = it.Next() {
 					val, isNull, err := baseFunc.evalDuration(row)
 					c.Assert(err, IsNil)
 					c.Assert(isNull, Equals, output.IsNull(i))
 					if !isNull {
-						c.Assert(val, Equals, d64s[i])
+						c.Assert(val, Equals, output.GetDuration(i, types.UnspecifiedLength))
```
Comment (XuHuaiyu) [inline expression/bench_test.go]: s/ types. UnspecifiedLength/ types.DefaultFsp

File: expression/builtin_cast_vec.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalInt(b.ctx, input, buf); err != nil {
+		return err
+	}
+
+	result.ResizeDuration(n, false)
+	i64s := buf.Int64s()
+	ds := result.GoDurations()
+	for i := 0; i < n; i++ {
+		if buf.IsNull(i) {
```
Comment (XuHuaiyu) [inline expression/builtin_cast_vec.go]: Can MergeNulls be used instead of this?

File: expression/builtin_cast_vec.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+			result.SetNull(i, true)
+			continue
+		}
+		dur, err := types.NumberToDuration(i64s[i], int8(b.tp.Decimal))
+		if err != nil {
+			if types.ErrOverflow.Equal(err) {
+				err = b.ctx.GetSessionVars().StmtCtx.HandleOverflow(err, err)
```
Comment (XuHuaiyu) [inline expression/builtin_cast_vec.go]: Should the warning message be refined like [this](https://github.com/pingcap/tidb/blob/master/expression/builtin_cast.go#L1090).

**Comment by zz-jason** [comment]
to https://github.com/pingcap/tidb/issues/12058

**Comment by zz-jason** [comment]
@qw4990 any update?

**Comment by qw4990** [comment]
PTAL @XuHuaiyu

File: expression/builtin_cast_vec.go (hunk 2 above)
Comment (zz-jason) [reply expression/builtin_cast_vec.go]: For this case, we need to copy the null bitmap from `buf` to `result`. I think we can add an interface to do it. @qw4990

Comment (qw4990) [reply expression/builtin_cast_vec.go]: We already have such an interface `MergeNulls` and it is used here now.

File: expression/builtin_cast_vec.go (hunk 3 above)
Comment (qw4990) [reply expression/builtin_cast_vec.go]: In its row-based implementation, this error is not refined.

**Pull request merged**